contracts/rateproducer/rateproducer.cpp

```diff
@@ -4,257 +4,470 @@
+   constexpr name system_account{"eosio"_n};
+   
+    struct producer_info {
+      name                  owner;
+      double                total_votes = 0;
+      eosio::public_key     producer_key; /// a packed public key object
+      bool                  is_active = true;
+      std::string           url;
+      uint32_t              unpaid_blocks = 0;
+      time_point            last_claim_time;
+      uint16_t              location = 0;
+
+      uint64_t primary_key()const { return owner.value;                             }
+      double   by_votes()const    { return is_active ? -total_votes : total_votes;  }
+      bool     active()const      { return is_active;                               }
+      void     deactivate()       { producer_key = public_key(); is_active = false; }
+
+      // explicit serialization macro is not necessary, used here only to improve compilation time
+      EOSLIB_SERIALIZE( producer_info, (owner)(total_votes)(producer_key)(is_active)(url)
+                        (unpaid_blocks)(last_claim_time)(location) )
+    };
+    
+     typedef eosio::multi_index< "producers"_n, producer_info,
+                               indexed_by<"prototalvote"_n, const_mem_fun<producer_info, double, &producer_info::by_votes>  >
+                             > producers_table;
+    
+    bool is_blockproducer(name bp_name){
+        producers_table bp(system_account, system_account.value);
+        auto it = bp.find(bp_name.value);
+        if(it==bp.end()){
+            return false;
+        }
+        return true;
+    }
 
-CONTRACT rateproducer : public contract {
-  public:
-    using contract::contract;
-    typedef struct bp_rate_t {
-      float transparency;
-      float infrastructure;
-      float trustiness;
-      float community;
-      float development;
-    } bp_rate_stats ;
 
-    ACTION rate(name user, name bp, string ratings_json) {
-      require_auth(user);
+   struct voter_info {
+      name                owner;
+      name                proxy;
+      std::vector<name>   producers;
+      int64_t             staked = 0;
+      double              last_vote_weight = 0;
+      double              proxied_vote_weight= 0;
+      bool                is_proxy = 0;
+      uint32_t            flags1 = 0;
+      uint32_t            reserved2 = 0;
+      eosio::asset        reserved3;
 
-      //TODO: bp must be a registered block producer
+      uint64_t primary_key()const { return owner.value; }
 
-      // the payload must be ratings_json.
-      check(ratings_json[0] == '{', "payload must be ratings_json");
-      check(ratings_json[ratings_json.size()-1] == '}', "payload must be ratings_json");
+      EOSLIB_SERIALIZE( voter_info, (owner)(proxy)(producers)(staked)(last_vote_weight)(proxied_vote_weight)(is_proxy)(flags1)(reserved2)(reserved3) )
+   };
 
-      // upsert bp rating
-      producers_table bps(_self, _self.value);
-      auto uniq_rating = (static_cast<uint128_t>(user.value) << 64) | bp.value;
```

Comment:
@kriskoin  where we not using this attribute  `uniq_rating` anywhere? 

We will need to query this table for records w

Reply:
@xavier506  
uniq_rating is composed key = voter name + bp name
is used everytime a vote is made
```
    producers_table bps(_self, _self.value);
    auto uniq_rating = (static_cast<uint128_t>(user.value) << 64) | bp.value;
    auto uniq_rating_index = bps.get_index<name("uniqrating")>();
    auto existing_rating = uniq_rating_index.find(uniq_rating);  < --- H E R E  ! ! !
```